### PR TITLE
cli: print commits abandoned by git::import_refs() or fetch()

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Review by martinvonz — APPROVED** [review]
Thanks! This seems useful.

By the way, one advantage of having `jj git sync` command is that it can avoid showing some of the conflicted commits to the user.

File: cli/src/cli_util.rs
Comment (ilyagr) [inline cli/src/cli_util.rs:1791]: Perhaps make this say "Import of git refs abandoned the following commits:" for the `jj git import` case? Otherwise, the user will think that `jj log` abandoned their commits.

File: cli/tests/test_git_push.rs
Comment (ilyagr) [inline cli/tests/test_git_push.rs:577]: Not sure if this is easily fixable, but I'm assuming that the commits are abandoned by the `git import` that happens *before* the push. It's slightly confusing that "Abandoned the following commits" is printed after the message about the push.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:1791]: For the fetch, you could perhaps also adjust the message to "Fetch of git refs...".

File: cli/tests/test_git_push.rs
Comment (yuja) [reply cli/tests/test_git_push.rs:577]: This one is branch deletion at remote, so the commit is abandoned after the push.

File: cli/src/cli_util.rs
Comment (yuja) [reply cli/src/cli_util.rs:1791]: I can't think of a good wording that works for all of the import/fetch/push cases. The automated import_refs() is confusing, but for the other cases, user should know the context.

So I think we can insert additional message if automated import_refs() detected changes.
```
Imported changes from underlying Git repo
Abandoned the following commits:
  kkmpptxz 35605592 master | (empty) bar
Rebased ... descendant commits
Working copy now at: yqosqzyt 096dc80d (empty) (no description set)
Parent commit      : qpvuntsm 230dd059 (empty) (no description set)
...
```

Comment (ilyagr) [reply cli/src/cli_util.rs:1791]: Perhaps you could pass a slightly different string to the function at each call site?

>  I think we can insert additional message if automated import_refs() detected changes.

This is an interesting idea, worth trying. It's not immediately clear to me whether this would be so noisy as to be annoying, but there's a good chance it's fine.

File: cli/tests/test_git_push.rs
Comment (ilyagr) [reply cli/tests/test_git_push.rs:577]: Ah, makes sense.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:1791]: Though, at a glance I don't see a separate call site for `git push`. :(

Comment (yuja) [reply cli/src/cli_util.rs:1791]: > insert additional message if automated import_refs() detected changes.

I've made this change. PTAL.

> pass a slightly different string to the function at each call site

It'll get complicated if we add other stats like new/changed local branches.

File: cli/src/cli_util.rs
Comment (ilyagr) [inline cli/src/cli_util.rs:799]: If we go with this, there should be a period after "repo". Also, perhaps this should be labeled so that we can make it brightblack or some other color?

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:1791]: > I've made this change. PTAL.

I'm still worried about this being annoying. I can try using it for a while and see how bad it gets.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:799]: Another idea: perhaps put this after the messages about abandoned commits? The message could (optionally) be changed to emphasize the order of operations:

"Done importing changes from the underlying Git repo."

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:1791]: I'm also still worried about the `push` case. I'm not sure how to make that any clearer without customizing the message for the call site. Fortunately, users shouldn't see that too often.

I didn't follow your thought about how complicated it will get. Superficially, it seems like we're OK now and I hope becoming more vague if necessary in the future might be good enough. But if you think that customization is not an option, we can try this way and see how it goes.

File: cli/src/cli_util.rs
Comment (ilyagr) [reply cli/src/cli_util.rs:799]: My main motivation is to make it *really* clear that something like `jj describe` or `jj log` is not the reason commits got abandoned.

Comment (yuja) [reply cli/src/cli_util.rs]: Sounds good. Moved the message to the bottom.

File: cli/src/cli_util.rs
Comment (yuja) [reply cli/src/cli_util.rs:1808]: > I'm still worried about this being annoying.

It'll be showed only when you've changed the repo by git command, so I hope it wouldn't be too verbose. We have some test changes where external git command isn't involved, but they're probably a bug of `git::export_refs()`.

> I'm also still worried about the push case.

I think the push scenario is fundamentally unintuitive. The user might just want to push the deleted ref (without abandoning the local commit.)

Comment (ilyagr) [reply cli/src/cli_util.rs:1808]: > It'll be showed only when you've changed the repo by git command, so I hope it wouldn't be too verbose. 

Yes, I tried it and it's much better than I thought it would be.

I also noticed that this kind of output is easy to miss for me because I have an analogue of `watch jj log` running.

> We have some test changes where external git command isn't involved, but they're probably a bug of git::export_refs().

Interesting.

**Review by ilyagr — APPROVED** [review]
I'm still a bit uneasy about this (whether it will create confusion, especially for beginners that didn't notice the abandoned commits and don't quite understand what it means), but this is clearly useful for people who notice the abandoned commits themselves. I think there's a good chance it's fine. We should try this version and see how (and whether) people react.

Yuya mentioned something intersting about this triggering unexpected test changes. I'll leave it up to him whether to look into it before or after merging this.

**Comment by ilyagr** [comment]
Here's the first time this PR became really noticeable. Most (all?) of these branches were moved, not deleted. I can't yet decide whether I like it.

image

**Pull request merged**
